### PR TITLE
[MM-49921] Cloud export compliance

### PR DESCRIPTION
**Pull request by nickmisasi**
#### Summary
Adds a new generic AddressForm component that powers entering a shipping address in the payment modal. Upon receiving a compliance failed code of 422 from the backend while upgrading plans, a purchase pending modal will be displayed. Subsequent visits to the purchase modal will display that same pending screen, until the compliance screen has been manually acted upon on our side.

image
image
image

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-49921

#### Related Pull Requests

- Has server changes (please link here)
- Has mobile changes (please link here)

#### Release Note

```release-note
none
```

### CODE REVIEW

**Comment by johndavidlugtu** [comment]
Good stuff @nickmisasi -- let's re-use the copy from here: https://mattermost.atlassian.net/browse/MM-49496?focusedCommentId=140976 to keep things consistent

File: components/purchase_modal/purchase_modal.tsx
Comment (neallred) [inline components/purchase_modal/purchase_modal.tsx:1172]: Nit: I think we can avoid wrapping this in Stripe's `<Elements>` provider for this branch. It will avoid some unnecessary work in the case that they come back to this later while still in screening. I think the loadStripe block up above can also be skipped if they are in screening.

File: components/payment_form/address_form.tsx
Comment (neallred) [inline components/payment_form/address_form.tsx:49]: This requires a dom attribute matching the internal type name, which is tricky to ensure it is right at first and easy to accidentally break in subsequent changes. Looks like there's a mismatch between `line` vs `address` 1 & 2 at the moment. Could we make this more explicit with something like the following?

```tsx
    const handleInputChange = (key: keyof Address) => (
        event:
        | React.ChangeEvent<HTMLInputElement>
        | React.ChangeEvent<HTMLSelectElement>,
    ) => {
        const newStateValue = {
           [key]: value,
        }
        ...
     }
  
  ...
  <Input
                    name='address'
                    type='text'
                    value={address.line1}
                    onChange={handleInputChange('line1')}
```

File: components/purchase_modal/purchase_modal.tsx
Comment (neallred) [inline components/purchase_modal/purchase_modal.tsx:1148]: Could we make a helper to create a full blank address? There's some issues with the type coercion:
* the underlying inputs that receive an `undefined` at first will change from being "uncontrolled" to "controlled" inputs as they get filled in. This goes against what React wants and creates console errors.
* It can be hard to ensure all the places that trust and assume the fields match the types won't explode or have unexpected behavior. AFAICT it works out at the moment, but its another place that lends itself to hard to detect regressions.

File: components/payment_form/address_form.tsx
Comment (neallred) [inline components/payment_form/address_form.tsx:54]: Nit: We can skip the onAddressChange null check since this is a required prop, and this component can't do anything meaningful without it.

File: components/payment_form/address_form.tsx
Comment (neallred) [inline components/payment_form/address_form.tsx:27]: I've seen this pattern elsewhere, especially, in CWS, of taking in initial state from the parent, making a local copy, maintaining the copy, and sending changes out to the parent. It seems indirect and hard to tell what the source of truth is to me.

This works as is so its not a blocking change, but I think it would be simpler and easier to understand to pass in the parent state and always read it, and rely on the parent handler to update the state that will be sent back in. We always send in part of the parent state already in the form of the partial `{country: ...}` object. Always reading the parent would help us be more meticulous in the parent about sending in a fully fleshed out object instead of a type coerced object, and would let us avoid the `| null` in the parent to begin with.

**Review by neallred — CHANGES_REQUESTED** [review]
LGTM overall! The blocking things to me are:
* Having a smaller svg image for the screening flow
* looks like with the `subscribeCloudSubscription` parameters changes, some of the existing usages need to be updated to fix type check errors
* I think we want to say screening will be finished in 3 days instead of 1.

File: components/common/svg_images_components/compliance_screen_failed_svg.tsx
Comment (nickmisasi) [reply components/common/svg_images_components/compliance_screen_failed_svg.tsx:1]: I grabbed this from the Figma designs for this same flow on the portal side. I too thought it was way bigger than it should be but forgot to go back to it. Will take a look

File: components/purchase_modal/process_payment_setup.tsx
Comment (M-ZubairAhmed) [inline components/purchase_modal/process_payment_setup.tsx:339]: wondering if `t` should be used here instead of this.props.intl ?

Comment (neallred) [reply components/purchase_modal/process_payment_setup.tsx:339]: The `IconMessage` component supports two apis. The API being used asks the user provide only an id, and the IconMessage component renders it as a FormattedMessage (without providing a defaultMessage). This is how a lot of the existing usage is.

There is also a formatted<PropName> api, that requests the user pass in an already internationalized ReactNode. In general I think the component should move towards removing all usages of the id-only API because of the pitfall of not providing a defaultMessage: languages not providing the translation string will make errors.

That said, there are other existing usages of the id-only api so I'm 0/5 on whether we want to make this usage use the formatted* API in this PR.

**Comment by annelieseklein** [comment]
Looks great!

1. @nickmisasi here's an example in terms of size of illustration and width of body copy text https://www.figma.com/file/Aob5ffHtfjKzQoebR93ycA/Cloud-Freemium?node-id=508%3A436984&t=MqtksV0Hnsp90QMH-1
2. the illustration you have seems to have her arm cut out - can we fix that? let me know if you need a new illustration
3. see here for how the checkbox item should look like in terms of sizing and font: https://www.figma.com/file/HU7OqEjUyvfqe5vOY0Ice7/Untitled?node-id=0%3A35228&t=ZSfhNAzuocPpZJ1k-1

File: components/purchase_modal/process_payment_setup.tsx
Comment (M-ZubairAhmed) [reply components/purchase_modal/process_payment_setup.tsx:339]: Yup no worries at all just wanted to double check its usage

File: i18n/en.json
Comment (nickmisasi) [reply i18n/en.json:307]: Wrote this back in August, when it _was_ 1 business day 😂 fixed

File: components/purchase_modal/purchase_modal.tsx
Comment (nickmisasi) [reply components/purchase_modal/purchase_modal.tsx:1148]: Added a helper. I always forget Typescript isn't like Go where it will create an entire struct including all fields with zero values

File: components/payment_form/address_form.tsx
Comment (nickmisasi) [reply components/payment_form/address_form.tsx:27]: I've updated the whole component to make it controlled. Did it the way it was initially because I basically just pulled the form out of the `payment_form` and adjusted it

**Comment by nickmisasi** [comment]
@annelieseklein 

I've set a max-width of 560px on the IconMessage component. Note that this will take effect for all occurrences of this component like the upgrade success/failure screen, downgrade success/failure screen, etc.

image


I used this illustration https://www.figma.com/file/GtAkFsG8unkQkcIudmAJwE/Export-Compliance?node-id=65%3A1346&t=NDQmUeVCjX0LMWqe-4 which is what we currently have in the portal. If you'd like something else just shoot me a link 

I'll get the checkbox styling cleaned up and get back to you

**Comment by nickmisasi** [comment]
@neallred I updated the subscribeCloudSubscription to optionally take the shipping address since downgrades/monthly won't actually need to provide this information. 

Going to wait on Anneliese before looking into the SVG size 

Otherwise I think I've addressed everything

**Review by neallred — APPROVED** [review]
LGTM pending resolution of the SVG question. Thanks, @nickmisasi !

File: components/purchase_modal/index.ts
Comment (marianunez) [inline components/purchase_modal/index.ts:52]: Is there a related CWS PR that includes this new field?

File: i18n/en.json
Comment (johndavidlugtu) [inline i18n/en.json:307]: > Good stuff @nickmisasi -- let's re-use the copy from here: https://mattermost.atlassian.net/browse/MM-49496?focusedCommentId=140976 to keep things consistent

Bringing this here so it's in-context @nickmisasi

Comment (nickmisasi) [reply i18n/en.json:307]: `Your transaction is being reviewed`

`We will check things on our side and get back to you within 3 days once your Cloud Subscription upgrade is approved. In the meantime, please feel free to continue using the free version of our product.`

What you linked mentions licenses. Does `Cloud Subscription upgrade` in place of `license` work?

Is the CTA ok?

Comment (johndavidlugtu) [reply i18n/en.json]: * Cloud subscription upgrade

CTA: Continue with Cloud Free

Comment (nickmisasi) [reply i18n/en.json]: @johndavidlugtu all set
image

File: components/purchase_modal/index.ts
Comment (nickmisasi) [reply components/purchase_modal/index.ts:52]: I thought I had requested review on them but I guess I did not. Should be there now.

**Review by johndavidlugtu — APPROVED** [review]
Awesome, thanks

**Pull request merged**
